ledger-tool/src/main.rs

```diff
@@ -2350,6 +2482,72 @@ fn main() {
                         system_monitor_service.join().unwrap();
                     }
                 }
+                ("simulate-block-production", Some(arg_matches)) => {
+                    let marker_file = simulate_block_production_marker_path(&ledger_path);
+                    if !marker_file.exists() {
+                        eprintln!(
```

Comment:
Could we re-use the blockstore primary/secondary usage detection for this purpose? I think that mainly helps us prevent use on (at least running) production systems.

Reply:
hehe, i know this is opinionated.

but, i also want to not-running production systems and not-always-running ledgers lying on dev boxes.

Reply:
Hmm, I think I agree with Andrew on using the existing check, and not introducing this additional check. The `is_primary()` check isn't perfect, but at least we're consistent in its' use.

Also, the intention of the marker file is good, but I think it is somewhat flawed. Namely, you might create the marker file during one session, finish your session and then leave the file in your ledger. Then, next time you go to run the command, the marker file is still there (so you don't get the check this second session).

Reply:
> Also, the intention of the marker file is good, but I think it is somewhat flawed. Namely, you might create the marker file during one session, finish your session and then leave the file in your ledger. Then, next time you go to run the command, the marker file is still there (so you don't get the check this second session).

Actually, leaving the file is an intended part. As `simulate-block-production` purges and writes somewhat arbitrary slot as if _real_, i wanted to mark the ledger as _contaminated_ forever, unless explicitly the marker file is removed.

Intended session of this command use is that you first create a small ledger copied from some dev ledger or production ledger. this command should never run on those original ledgers. So, the marker file functions as a chilling effect as one is now forced to manually create some random file by `touch`-ing into those ledgers.

Reply:
> i wanted to mark the ledger as contaminated forever, unless explicitly the marker file is removed ...

Ok, I see what you're going for. I still have concerns about visibility of this marker though. I hope nobody would run this on one of our prod mnb/testnet/devnet nodes, but when I log into any of those for miscellaneous debugging, I don't necessarily go and run `ls ledger`; usually my first command is `bounds` and go from there.

Reply:
> I hope nobody would run this on one of our prod mnb/testnet/devnet nodes

me too. :trollface:

> I still have concerns about visibility of this marker though. ... I don't necessarily go and run ls ledger; 

if you feel strong about this, i can make validator abort on finding the marker file :grin:

Reply:
> if you feel strong about this, i can make validator abort on finding the marker file 😁

On the contrary, I'm still leaning more towards removing the marker file + check for it.

Here is another example scenario:
- User runs `simulate-block-production` on some ledger
- User then runs `blockstore copy` to copy simulated shreds to some other blockstore

In this case, we are moving "contaminated" blocks over without the "contaminated" marker. I don't think we should copy the marker over for this case; rather, I think we can proceed with some expectation that the user is running a command and knows what they're doing. As you mentioned, this is the intended workflow:

> Intended session of this command use is that you first create a small ledger copied from some dev ledger or production ledger. this command should never run on those original ledgers

In general, I think protections are good, but if the protection isn't comprehensive, then someone might let their guard down because they think the tool will save them from their errors.

Lastly, I don't love the one-off nature of the file / check. There are other commands that can alter the ledger, but those don't require a marker file. Maybe they should have some "opt-in" to allow it, but if so, I think the opt-in should be done in a way that is consistent for all command.

Reply:
thanks for detailing your concerns. I just removed the marker file and check: f4a1ca8ea3d

Reply:
oops, i also needed to remove this as well: https://github.com/anza-xyz/agave/pull/2733/commits/0e8aca5fdbcf201620583d87c7b98363865a57b3